### ISSUE TITLE
Add ScriptLoad hook

### ISSUE THREAD

**Issue by Tripperful** [Facepunch/garrysmod-requests#1467]
This hook should be called every time any lua script is loaded/refreshed, not talking about RunString, just the regular lua files.

Script file path should be passed as an argument, and if the hook returns a string, this string is used as script content, instead of script file content.

What would this enable us to do:

- **Create and use Lua preprocessors/transpilers in a less hacky way**
- See which scripts get loaded and when
- Block certain scripts from loading by returning an empty string instead of their content
- Many more useful things

**Comment by ExtReMLapin** [comment]
Oof, some drm guys gonna be pissed it it gets released.


Hook all function run

if function like == -1 then it means it's a file that you run

If the file path seems suspicious it means it's runstring

dump bytecode/change values in it and boom

**Comment by Tripperful** [comment]
@ExtReMLapin yes, we can already do this, but I think you understand what I mean saying "in a less hacky way"
If any "drm guys" wont be happy, that's just a little bonus

**Comment by ExtReMLapin** [comment]
It's not a gmod hook, you can't really prevent a script from loading but you can hook files being loaded (after they were compiled to lua bytecode and called)

https://github.com/ExtReMLapin/BadCoderz/blob/master/lua/badcoderz/sh_compiled_functions.lua#L52


After the selected line on the link, it's code related to find runstring/compilestring code, modify it to fits your needs

**Comment by brandonsturgeon** [comment]
This would be nice for a number of reasons, especially if it can simply be a refresh hook (regardless of if the file actually _ran_ after being refreshed)

For example:
```lua
hook.Add( "OnLuaFileRefresh", "test", function( filePath )
    print( "Refresh detected", filePath )
end )
```

This could allow dynamic for-loop `include` loaders to detect when one of their dynamic files was refreshed, and manually `include` it where the standard refresh system wouldn't know better.
